controllers/resources/lease.go

```diff
@@ -0,0 +1,263 @@
+	UpdateReconcileResults(ctx context.Context, nhc *remediationv1alpha1.NodeHealthCheck, initialRequeue time.Duration, initialErr error) (time.Duration, error)
+}
+
+type nhcLeaseManager struct {
+	client             client.Client
+	commonLeaseManager lease.Manager
+	log                logr.Logger
+}
+
+func NewLeaseManager(client client.Client, log logr.Logger) (LeaseManager, error) {
+	newManager, err := lease.NewManager(client, holderIdentity)
+	if err != nil {
+		log.Error(err, "couldn't initialize lease manager")
+		return nil, err
+	}
+	return &nhcLeaseManager{
+		client:             client,
+		commonLeaseManager: newManager,
+		log:                log.WithName("nhc lease manager"),
+	}, nil
+}
+
+func (m *nhcLeaseManager) ObtainNodeLease(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) (bool, *time.Duration, error) {
+	nodeName := remediationCR.GetName()
+	leaseDuration := m.calculateLeaseDurationBySingleRemediation(remediationCR, nhc)
+	leaseDurationWithBuffer := leaseDuration + LeaseBuffer
+
+	node := &v1.Node{}
+	if err := m.client.Get(context.Background(), types.NamespacedName{Name: nodeName}, node); err != nil {
+		m.log.Error(err, "couldn't obtain node lease node error getting node", "node name", nodeName)
+		return false, nil, err
+	}
+
+	if err := m.commonLeaseManager.RequestLease(context.Background(), node, leaseDurationWithBuffer); err != nil {
+		if _, ok := err.(*lease.AlreadyHeldError); ok {
+			m.log.Info("can't acquire node lease, it is already owned by another owner", "already held error", err)
+			return false, &RequeueIfLeaseTaken, err
+		}
+
+		m.log.Error(err, "couldn't obtain lease for node", "node name", nodeName)
+		return false, nil, err
+	}
+
+	//all good lease created with wanted duration
+	return true, &leaseDuration, nil
+
+}
+
+func (m *nhcLeaseManager) UpdateReconcileResults(ctx context.Context, nhc *remediationv1alpha1.NodeHealthCheck, initialRequeue time.Duration, initialErr error) (time.Duration, error) {
+	originalRequeue, originalErr := initialRequeue, initialErr
+	updatedRequeue, updatedErr := initialRequeue, initialErr
+	if requeue, err := m.manageLeases(ctx, nhc); err != nil {
+		if initialErr == nil {
+			updatedErr = err
+		} else {
+			updatedErr = pkgerrors.Wrap(initialErr, err.Error())
+		}
+		//requeue returned from lease manager is lower than the original request requeue
+	} else if requeue > 0 && (initialRequeue == 0 || requeue < initialRequeue) {
+		updatedRequeue = requeue
+	}
+	//Log
+	m.logManageLeaseChanges(originalRequeue, updatedRequeue, originalErr, updatedErr)
+	return updatedRequeue, updatedErr
+}
+
+func (m *nhcLeaseManager) calculateLeaseDurationBySingleRemediation(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) time.Duration {
```

Comment:
>would getTimeoutForRemediation() be a more descriptive name?

Sure, that's fine by me

>And use the default at caller side when this returns 0?

I actually rather keep this in the current method, I think extracting it will cause less clean code. 
(I'll even remove the redundant else in `calculateLeaseDurationBySingleRemediation` :smile: )

Reply:
It's about separation of concerns, or modularity, or whatever it's called. And about meaningful names.
The thing the method is doing is: find the timeout. That's why I proposed the new name. But then the default lease duration has nothing to do with the timeout.
We could add an intermediate method:
```
getLeaseDurationForRemediation() {
  if  timeout := getTimeoutForRemediation(); timeout != 0 {
     return timeout
  }
  return defaultLeaseDuration
}
```
But IMHO this isn't worth it, it can be handled by the caller of getTimeoutForRemediation() directly.
Does that make sense? 🤔